client/asset/interface.go

```diff
@@ -814,6 +822,37 @@ type BotWallet interface {
 	SingleLotRedeemFees(*PreRedeemForm) (uint64, error)
 }
 
+// ShieldedBalance is the balance associated with a ZCash shielded address.
+// Unlike regular utxo accounts, a shielded address is supposed to be reused,
+// making a shielded address more like an "account".
+// https://electriccoin.co/blog/shielded-address-contexts/
+type ShieldedBalance struct {
+	Address string
+	Balance uint64
+}
+
+// ShieldedWallet is implemented by ZCash, and enables working with value in
+// shielded pools.
+type ShieldedWallet interface {
+	// ShieldedBalances list the address-balance pairs for the wallet's shielded
+	// account.
+	ShieldedBalances() ([]*ShieldedBalance, error)
+	// NewShieldedAddress creates a new shielded address. Unlike regular utxo
+	// assets, you do not create a shielded address for every transaction. A
+	// shielded address is to be reused, so creating a new address is more like
+	// creating a new account.
+	NewShieldedAddress() (string, error)
+	// ShieldFunds moves funds from the transparent account to the specified
+	// shielded address.
+	ShieldFunds(ctx context.Context, shieldedAddr string, amt uint64) (dex.Bytes, error)
+	// UnshieldFunds moves funds from the specified shielded address to the
+	// transparent account.
+	UnshieldFunds(ctx context.Context, shieldedAddr string, amt uint64) (dex.Bytes, error)
+	// SendShielded sends funds from the specified shielded account to the
+	// provided shielded or transparent address.
+	SendShielded(ctx context.Context, shieldedAddr, toAddr string, amt uint64) (dex.Bytes, error)
```

Comment:
Since the `Core` changes come next, could you describe the intended flow of funds from deposits and trades into and out of shielded addresses, to effectively achieve mixing presumably?

Reply:
> Since the Core changes come next, could you describe the intended flow of funds from deposits and trades into and out of shielded addresses, to effectively achieve mixing presumably?

Great question! In the lingo I'm trying to normalize, the model goes as such. Every wallet has 2 parts. There's the transparent part, which is the only part that we've known so far. And there's the shielded part = the new secret part. All swap activity will take place on the transparent part. That's why the server will never be involved with this effort. But we can also transfer value into and out of the shielded part (on-chain). Transferring value from the transparent part to the shielded part does not reveal the shielded address. There is no way to view the balance of a shielded address by viewing blockchain data only. Tranferring value from, say, the Orchard shielded pool to another address in the Orchard shielded pool does not reveal the sender, receiver, or amount. This is the highest-security type of value transfer. Any other type of transfer, e.g. Orchard-to-Sapling, Orchard-to-transparent, transparent-to-Sapling, etc. will always reveal the transfer amount, which itself could potentially be used to reduce your anonymity. Any value transfer from or to a transparent address further reduces the anonymity, by revealing a pubkey hash (btc-like security).

Since the shielded address is never parseable from on-chain data, the shielded address can be reused (with some security implications for meat-world collaboration). For this reason, a shielded address is more akin to an "account". We don't need to generate a new address for every shielded transaction. We **can** generate new shielded addresses=accounts at will, but **we can't send value from more that one shielded address=account in a single transaction**. The address=account paradigm has serious implications for UI.


> And also a comment about why the consumer must be aware of this instead of the wallet handling everything itself, with things just being immature for a while, like with DCR.

With DCR in mixing mode, the wallet is assumed to be a properly configured remote dcrwallet instance accessed via RPC. It requires substantial external setup, but is expected to be self-sustaining after that. (When we implement Decred mixing natively to DEX, we will also need to expose more information to the consumer and to add UI elements.)

With ZCash, the situation is quite different, just because of the difference in the models. There is no "automatic" mode. Any use of privacy will require explicit interaction from the user.

Reply:
When you say all of the swap activity takes place in the transparent part, do you mean that someone has to unshield manually, do a swap, and then reshield manually? I think creating the swap transaction could itself be an unsheild, and the redeem could be a reshield transaction.

Reply:
I'm not opposed to a shielded-by-default mode, but I don't think it's right for a first approach. It would require modifying virtually every step of the trade process and a deeper integration into the BTC package. It would not be a small undertaking. It's also not what I promised in the prop, so any time put into such an effort would potentially detract from our ability to deliver on other promises.

Reply:
Wouldn't the only difference be to change the Swap to get inputs from the shielded pool and the redeem output to be to a shielded pool? Can definitely do it later though if there's time.

Reply:
Ah I guess proving you have control of funds to the server would be a challenge.